packages/engine-core/src/library/ExitHooks.ts

```diff
@@ -0,0 +1,67 @@
+import Debug from '@prisma/debug'
+
+export type BeforeExitListener = () => unknown
+
+const debug = Debug('prisma:client:libraryEngine:exitHooks')
+
+export class ExitHooks {
+  private nextOwnerId = 1
+  private ownerToIdMap = new WeakMap<object, number>()
+  private idToListenerMap = new Map<number, BeforeExitListener>()
+  private areHooksInstalled = false
+
+  install() {
+    if (this.areHooksInstalled) {
+      return
+    }
+
+    this.installHook('beforeExit')
+    this.installHook('exit')
+    this.installHook('SIGINT', true)
+    this.installHook('SIGUSR2', true)
```

Comment:
TODO: check whether support for the `SIGUSR2` signal can be dropped

Reply:
I guess not until prisma 5? Since `beforeExit` event is a part of [public API](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-management#exit-hooks)

Reply:
If we want to remove it in next major, let's mark it deprecated everywhere, then?

Reply:
Will do a proposal on deprecating it in 4.2.0 and removing in 5.0